File: native/chu/tasoller_v2.go
package chu

type TasollerV2Parser struct{}

func (p *TasollerV2Parser) GetName() string {
	return "TasollerV2"
}

func (p *TasollerV2Parser) Parse(raw []byte) ChusanData {
	if len(raw) < 36 {
		return ChusanData{}
	}

	data := ChusanData{}

	irButtons := raw[3]

	rawBtn := irButtons >> 6

	data.Test = (rawBtn&2)>>1 == 1
	data.Service = (rawBtn&1)<<1>>1 == 1

	beams := irButtons & 0x3F
	for i := 0; i < 6; i++ {
		data.Air[i] = (beams >> i) & 1 == 1
	}

	sliderRaw := raw[4:36]
	for i := 0; i < 32; i++ {
		var targetIdx int
		if i%2 == 0 {
			targetIdx = 30 - i
		} else {
			targetIdx = 32 - i
		}
		if targetIdx >= 0 && targetIdx < 32 {
			data.Slider[targetIdx] = sliderRaw[i]
		}
	}

	return data
}

File: native/src/chu/tasoller_v2.rs
use super::{ChusanData, ChusanParser};

pub struct TasollerV2Parser;

impl ChusanParser for TasollerV2Parser {
    fn get_name(&self) -> &'static str {
        "TasollerV2"
    }

    fn parse(&self, raw: &[u8]) -> ChusanData {
        if raw.len() < 36 {
            return ChusanData::default();
        }

        let mut data = ChusanData::default();

        let ir_buttons = raw[3];
        let raw_btn = ir_buttons >> 6;

        data.test = ((raw_btn & 2) >> 1) == 1;
        data.service = ((raw_btn & 1) << 1 >> 1) == 1;

        let beams = ir_buttons & 0x3F;
        for i in 0..6 {
            data.air[i] = ((beams >> i) & 1) == 1;
        }

        // Slider bytes with index swapping: even→30-i, odd→32-i
        let slider_raw = &raw[4..36];
        for i in 0..32 {
            let target = if i % 2 == 0 { 30 - i } else { 32 - i };
            if target < 32 {
                data.slider[target] = slider_raw[i];
            }
        }

        data
    }
}
